fast64_internal/f3d/f3d_material.py

```diff
@@ -1098,158 +1091,226 @@ def ui_procAnimField(procAnimField, layout, name):
+    "LOD_FRACTION": ("CombinerInputs", "LOD Fraction"),
+    "PRIM_LOD_FRAC": ("CombinerInputs", "Prim LOD Fraction"),
+    "NOISE": ("CombinerInputs", "Noise"),
+    "K4": ("CombinerInputs", "YUVConvert K4"),
+    "K5": ("CombinerInputs", "YUVConvert K5"),
+    "1": ("CombinerInputs", "1"),
+    "0": (None, 0),
+}
+
+# Maps the alpha combiner input name to the corresponding node name and output name
+alpha_combiner_inputs = {
+    "COMBINED": (None, "Alpha"),
+    "TEXEL0": ("Tex0_I", "Alpha"),
+    "TEXEL1": ("Tex1_I", "Alpha"),
+    "PRIMITIVE": ("CombinerInputs", "Prim Alpha"),
+    "SHADE": ("Shade Color", "Alpha"),
+    "ENVIRONMENT": ("CombinerInputs", "Env Alpha"),
+    "LOD_FRACTION": ("CombinerInputs", "LOD Fraction"),
+    "PRIM_LOD_FRAC": ("CombinerInputs", "Prim LOD Fraction"),
+    "1": ("CombinerInputs", "1"),
+    "0": (None, 0),
+}
+
+
+def remove_first_link_if_exists(material: bpy.types.Material, links):  # TODO: (V5) add links type annotation
+    if len(links) > 0:
+        link = links[0]
+        material.node_tree.links.remove(link)
+
+
+def link_if_none_exist(
+    material: bpy.types.Material, fromOutput, toInput
+):  # TODO: (V5) add output/input type annotations
+    if len(fromOutput.links) == 0:
+        material.node_tree.links.new(fromOutput, toInput)
+
+
+def update_node_combiner(material, combinerInputs, cycleIndex):
     nodes = material.node_tree.nodes
-    combinerNode = nodes["Color Combiner Cycle " + str(cycleIndex) + " F3D v3"]
+
+    if cycleIndex == 1:
+        cycle_node = nodes["Cycle_1"]
+    else:
+        cycle_node = nodes["Cycle_2"]
+
     for i in range(8):
-        for link in combinerNode.inputs[i].links:
-            material.node_tree.links.remove(link)
-        if i in range(0, 4):
-            if combinerInputs[i] == "COMBINED" or combinerInputs[i] == "COMBINED_ALPHA":
-                if cycleIndex == 2:
-                    combiner1 = nodes["Color Combiner Cycle 1 F3D v3"]
-                    combiner2 = nodes["Color Combiner Cycle 2 F3D v3"]
-                    material.node_tree.links.new(
-                        combiner2.inputs[i], combiner1.outputs[0 if combinerInputs[i] == "COMBINED" else 1]
-                    )
-            else:
-                if cycleIndex == 2 and combinerInputs[i][:5] == "TEXEL":
-                    value = combinerInputs[i]
-                    combinerInput = value.replace("0", "1") if "0" in value else value.replace("1", "0")
-                else:
-                    combinerInput = combinerInputs[i]
-                combinerSocket = getSocketFromCombinerToNodeDictColor(nodes, f3dVer, combinerInput)
-                material.node_tree.links.new(combinerNode.inputs[i], combinerSocket)
+        combiner_input = combinerInputs[i]
+        if cycleIndex == 2 and "TEXEL" in combiner_input:
+            combiner_input = (
+                combiner_input.replace("0", "1") if "0" in combiner_input else combiner_input.replace("1", "0")
+            )
```

Comment:
its a tad more complex, could be `TEXEL0_ALPHA` or `TEXEL1_ALPHA`

Reply:
that's one more reason to replace this code with something clearer then lol because it's super not obvious what this does monkaS
something like
```suggestion
        if cycleIndex == 2:
            swaps_tex01 = {"TEXEL0":"TEXEL1", "TEXEL0_ALPHA":"TEXEL1_ALPHA", ...}
            combiner_input = swaps_tex01.get(combiner_input, combiner_input)
```
then?

Reply:
updated